include/dlaf/common/assert.h

```diff
@@ -0,0 +1,38 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+#include <exception>
+#include <iostream>
+
+#include "dlaf/common/utils.h"
+
+#define DLAF_TEST(category, condition, ...)                                                  \
+  if (!(condition)) {                                                                        \
+    std::cerr << "[" category "-ERROR] " << __FILE__ << " : " << DLAF_SOURCE_LOCATION << ":" \
+              << __LINE__ << "\n"                                                            \
+              << dlaf::common::concat(#condition, ##__VA_ARGS__) << std::endl;               \
+    std::terminate();                                                                        \
+  }
+
+#ifndef NDEBUG
+#define DLAF_ASSERT(condition, ...) DLAF_TEST("ASSERT", condition, ##__VA_ARGS__)
+#else
+#define DLAF_ASSERT(condition, ...)
+#endif
+
+#ifdef DLAF_ENABLE_AUDIT
+#define DLAF_AUDIT(condition, ...) DLAF_TEST("AUDIT", condition, ##__VA_ARGS__)
+#else
+#define DLAF_AUDIT(condition, ...)
+#endif
+
+#define DLAF_PRECONDITION(condition, ...) DLAF_TEST("PRECONDITION", condition, ##__VA_ARGS__)
```

Comment:
In C++20 contracts, preconditions are checks on the parameters of a function. DLAF_ASSERT/AUDIT can do that too. Do we need this?

Reply:
This macro is exactly for checking parameters of algorithms. It is clearly an assert-like thing, but as stated before in the three levels hierarchy, we want them to be always on.